### PR TITLE
[WIP/hack] Restore previous instance when exiting with

### PR DESCRIPTION
**Pull request by CyberShadow**
This fixes enemies getting stuck in corners in Daniel Remar's Violet Metal, however I'm 99.9% sure this is the wrong fix.

Explanation:

In the `scr_sabotcreate` script (in Violet Metal), which runs at the start of each level, there's this code:

```c
//Place start blocks
with (obj_block)
    instance_create(x,y,obj_startblock);

//Calculate crossings
for (i=0;i<320;i+=16){
    for (j=0;j<240;j+=16){
        if (place_free(i,j)){
            ......
```

The script (`scr_sabotcreate`) runs on the "Create" event of `obj_sabot`, of which there is one per room, thus making it the script that runs when entering a new room (level). So, the object the script runs "on" is the `obj_sabot` instance.

Now, the problem:

When exiting the `with (obj_block)`, the "current object" - instead of being restored to the initial `obj_sabot` instance - is incorrectly set to an `obj_startblock` instance. I don't know how exactly that happens, but it looks like the linked list node that `instance_event_iterator` points to is changed, so saving just the value of `instance_event_iterator` does us no good - it's the value that it's pointing at that changes.

Things continue to go south in the "Calculate crossings" code below: because the current instance is now no longer set to `obj_sabot`, it causes `place_free` to misbehave, as it's testing the collision mask of the wrong object. This leads the game to never place any (invisible) crossing tiles (`obj_crossing`), thus causing the AI to never "cross the street" and eventually end up and get stuck in corners.

This hack restores the `inst` in the linked list after also restoring the value of `instance_event_iterator` when exiting the `with` block, but it's almost certainly the wrong thing to do because it probably overwrites the `inst` in a linked list we shouldn't be writing to. I don't know who owns that linked list, though.

Anyway, I wanted to get this written down as I think I got as far as I can reasonably get without asking someone, and there's no one online on IRC right now to ask.

### CODE REVIEW

**Comment by JoshDreamland** [comment]
The actual fix is in [instance_system_base.h](https://github.com/enigma-dev/enigma-dev/blob/master/ENIGMAsystem/SHELL/Universal_System/instance_system_base.h#L39)/[instance_system.cpp](https://github.com/enigma-dev/enigma-dev/blob/master/ENIGMAsystem/SHELL/Universal_System/instance_system.cpp#L219). You'll notice that right now, the two are using a global singleton, `dummy_event_iterator`. That's just a sentinel; it shouldn't be using that. Instead, please have the type declare its own `inst_iter`, populate that, and take its address.

**Comment by JoshDreamland** [comment]
Not necessary, no. In fact, I explicitly recommend against doing that, as doing so could corrupt an actual event iterator in, say, a collision event, if the other instance was destroyed.

File: ENIGMAsystem/SHELL/Universal_System/instance_system.cpp [hunk 1]
```diff
@@ -216,9 +216,19 @@ namespace enigma
   inst_iter *instance_event_iterator = &dummy_event_iterator; // Not bad for efficiency, either.
   object_basic *instance_other = NULL;
 
-  temp_event_scope::temp_event_scope(object_basic* ninst): oinst(instance_event_iterator->inst), oiter(instance_event_iterator), prev_other(instance_other)
-    { instance_event_iterator = &dummy_event_iterator; instance_event_iterator->inst = ninst; instance_other=ninst; }
-  temp_event_scope::~temp_event_scope() { instance_event_iterator = oiter; instance_event_iterator->inst = oinst; instance_other=prev_other; }
+  temp_event_scope::temp_event_scope(object_basic* ninst) :
+    oiter(instance_event_iterator),
+    prev_other(instance_other),
+    iter(ninst, NULL, NULL)
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Universal_System/instance_system.cpp]: For future reference, we're trying to move to Google style, which places opening curlies on the same line as the statement or function prototype.

File: ENIGMAsystem/SHELL/Universal_System/instance_system_base.h [hunk 2]
```diff
@@ -38,9 +38,9 @@ namespace enigma
 
   class temp_event_scope
   {
-    object_basic *oinst;
     inst_iter *oiter;
     object_basic* prev_other; //Should always be NULL, but just in case...
```
Comment (JoshDreamland) [inline ENIGMAsystem/SHELL/Universal_System/instance_system_base.h]: Two more notes (no action required):
1. This comment of mine is wrong. This line of reasoning is the reason this bug exists. I think this mechanism was originally intended for instance creation scripts, and only later came to be used in regular creation events. All that needs to happen for this to not be `NULL` is for `instance_create` to be called in a collision event.
2. I believe we call the variable below `niter`, elsewhere (similar to `ninst`). This naming convention is largely unhelpful, though. Even just saying "new" instead would have been better....

**Comment by CyberShadow** [comment]
> This still fixes Violet, correct?

Yes, it does!

> If so, I will merge it, now.

I can fix the naming / style first if you like.

**Comment by JoshDreamland** [comment]
To be honest, I'm way more concerned about that completely wrong comment of mine. If you want to fix that, though, I'm certainly not opposed. ☺

**Comment by CyberShadow** [comment]
Fixed all the things. :)

**Comment by JoshDreamland** [comment]
Awesome; thank you!

**Comment by RobertBColton** [comment]
Got any specific test cases I can run to test this out? I am not interested in testing his game, since he already obviously tested it. I just pulled this and tested the FPS example plus this other AI tutorial for GMS that uses `with()` and didn't notice any regressions. I tested out some other random games and platforming examples from my "GMGames" folder too, no signs of regress.

**Comment by JoshDreamland** [comment]
This is one of those "certifiably correct" changes you get to make every once in a while, Robert. I'm going to add a regression test for this tonight, after the test harness is checked in.

**Comment by RobertBColton** [comment]
You don't even know it, but I'm really just that little voice of confirmation bias in the back of your head all day long telling you that you're right. I am the living breathing personification of confirmation bias in other words. Also, just so you know, Morty is eventually going to kill Rick C-137.

**Pull request merged**